crates/compiler/src/lib.rs

```diff
@@ -0,0 +1,746 @@
+use std::collections::HashMap;
+
+use ayysee_parser::ast::{Block, Expr, Identifier, Statement, Value};
+use stationeers_mips::{
+    instructions::{Arithmetic, FlowControl, Instruction, Misc, Stack as StackInstruction},
+    types::{Number, Register},
+};
+
+use crate::error::{Error, Result};
+
+pub mod error;
+
+#[derive(Copy, Clone, Debug)]
+/// The pass of the compiler.
+/// Multiple passes are used so that the compiler can resolve forward references.
+enum Pass {
+    /// The first pass, which resolves forward references.
+    /// Code will still be generated to reserve space for instructions that come later.
+    /// However, the second pass will overwrite these instructions with the correct values.
+    First,
+    Second,
+}
+
+#[derive(Copy, Clone, Debug)]
+#[allow(dead_code)]
+enum Location {
```

Comment:
This is intended to help the stack keep track of where function arguments are located